### PR TITLE
added underscored number restriction to close #205

### LINKED ISSUE #205: Restrict names with underscored numbers
# Rule request

## Thesis

There are two possible ways how one can write "Star Wars Episode 2" in snake case: `star_wars_episode2` and `star_wars_episode_2`.

The same goes for numbers in the middle of the name: `come2_me` vs `come_2_me`.

We need to enforce the first version and restrict the former.

## Reasoning

This is done for consistency in naming.

### PR DESCRIPTION
**Pull request by riyasyash**
# I have made things!
Closes  #205 

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [ ] I have added my changes to the `CHANGELOG.md`

### CODE REVIEW

File: wemake_python_styleguide/logics/filenames.py [hunk 1]
```diff
@@ -103,3 +103,22 @@ def is_matching_pattern(
...
+
+
+def is_stem_with_underscored_number(
+    file_path: str,
+    pattern: Pattern = constants.UNDERSCORED_NUMBER_PATTERN,
+) -> bool:
+    r"""
+    Checks whether the file's stem contains an underscored number.
+
+    >>> is_stem_with_underscored_number('episode2.py')
```
Comment (sobolevn) [inline wemake_python_styleguide/logics/filenames.py]: We use empty lines between doc tests.

File: wemake_python_styleguide/logics/filenames.py [hunk 2]
```diff
@@ -103,3 +103,22 @@ def is_matching_pattern(
     """
     stem = _get_stem(file_path)
     return pattern.match(stem) is not None
+
+
+def is_stem_with_underscored_number(
+    file_path: str,
+    pattern: Pattern = constants.UNDERSCORED_NUMBER_PATTERN,
+) -> bool:
+    r"""
```
Comment (sobolevn) [inline wemake_python_styleguide/logics/filenames.py]: `r` here is not required.

File: wemake_python_styleguide/logics/variables.py [hunk 3]
```diff
@@ -77,3 +78,29 @@ def is_private_variable(name: Optional[str]) -> bool:
     return (
         name is not None and name.startswith('__') and not name.endswith('__')
     )
+
+
+def is_variable_name_with_underscored_number(name: str) -> bool:
+    """
+    Checks for variable names with underscored number.
+
+    >>> is_variable_name_with_underscored_number('star_wars_episode2')
```
Comment (sobolevn) [inline wemake_python_styleguide/logics/variables.py]: The same about docs tests here.

File: wemake_python_styleguide/violations/naming.py [hunk 4]
```diff
@@ -439,3 +441,33 @@ class SameAliasImportViolation(ASTViolation):
...
+
+    Solution:
+        Rename your variable or modules to not include underscored numbers
+
+    This rule checks: variables, and modules.
+
+    Example::
+
+        # Correct:
+        star_wars_episode2
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/naming.py]: Can you please also list that `iso_123_456` is a valid name?

File: tests/test_visitors/test_filenames/test_wrong_module_name/test_module_magic_name.py [hunk 5]
```diff
@@ -31,3 +32,16 @@ def test_simple_filename(assert_errors, filename, default_options):
     visitor.run()
 
     assert_errors(visitor, [WrongModuleMagicNameViolation])
+
+
+@pytest.mark.parametrize('filename', [
```
Comment (sobolevn) [inline tests/test_visitors/test_filenames/test_wrong_module_name/test_module_magic_name.py]: Please, add happy path tests. I mean that we should test that `version2.py` and `version2_8.py` are correct names.

File: tests/test_visitors/test_ast/test_general/test_wrong_name/test_variable_names.py [hunk 6]
```diff
@@ -162,3 +163,23 @@ def test_correct_variable_name(
     visitor.run()
 
     assert_errors(visitor, [])
+
+
+@pytest.mark.parametrize('code', [
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_general/test_wrong_name/test_variable_names.py]: Please add happy path tests. We should test that variables like `test34` and `test32_109` are fine.

Comment (riyasyash) [reply tests/test_visitors/test_ast/test_general/test_wrong_name/test_variable_names.py]: Sure

File: tests/test_visitors/test_filenames/test_wrong_module_name/test_module_magic_name.py (hunk 5 above)
Comment (riyasyash) [reply tests/test_visitors/test_filenames/test_wrong_module_name/test_module_magic_name.py]: Sure

File: wemake_python_styleguide/logics/filenames.py (hunk 2 above)
Comment (riyasyash) [reply wemake_python_styleguide/logics/filenames.py]: Noted :)

File: wemake_python_styleguide/logics/filenames.py (hunk 1 above)
Comment (riyasyash) [reply wemake_python_styleguide/logics/filenames.py]: Done.

File: wemake_python_styleguide/logics/variables.py (hunk 3 above)
Comment (riyasyash) [reply wemake_python_styleguide/logics/variables.py]: Done.

File: wemake_python_styleguide/violations/naming.py (hunk 4 above)
Comment (riyasyash) [reply wemake_python_styleguide/violations/naming.py]: Added.

**Comment by sobolevn** [comment]
@riyasyash please pay attention to this issue: 

> /home/travis/build/wemake-services/wemake-python-styleguide/tests/test_visitors/test_ast/test_general/test_comparison_order_func.py:0:1: Z114 Found underscored name pattern "nonetype"

**Pull request merged**